RFC: netboot from all interfaces

Natolumin opened the pull request:
For #46 we want to be able to boot from all interfaces if any of them is configurable and will offer us dhcp.
See the commit message for the more detailed rationale.

This is RFC because I'm not sure this is the best way of doing it - the alternative I think is having netboot take only one interface as it is now, but making uinit or whatever else calls netboot do the interface discovery. However that would prevent parallelizing bringing up and waiting for the interfaces, for which we have a 30s timeout right now

Natolumin committed:
netboot: Attempt booting from all interfaces

The default of booting from eth0 doesn't work that well on devices that
have multiple interfaces, or would boot from non-ethernet interfaces.
This adds a mode where the available interfaces are autodetected and all
tried.

This can make the boot time longer on devices with multiple interfaces
(+30s per interface), so to mitigate it also lowers the interface wait
time. I don't think any reasonable interface would take 30s to boot or
even 10s; in the future we can make this timeout configurable

A way to reduce this wait time would be to start by initializing all
interfaces, then go over the interfaces as soon as they appear up

Natolumin commented on netboot/main.go:
```diff
@@ -62,134 +64,168 @@ func main() {
+			log.Fatal("Could not obtain interfaces from netlink; aborting")
+		}
+		for _, l := range links {
+			if l.Attrs().Name == "lo" {
```
Note this is bad and I should check for the flag LOOPBACK instead, as it doesn't have to be named lo - will update this before final version

insomniacslk replied:
yes, please. We can also add this as a separate function to the base library, as it would be used in other places too

insomniacslk replied:
first attempt at https://github.com/insomniacslk/dhcp/pull/217, cc @pmazzini

Natolumin replied:
I'm not sure this warrants a whole function, the abstraction provided by the netlink library is good enough for this to still be one line and explicit/readable. I just didn't think about it before sending the PR, it's not that it was more difficult than doing this

```go
l.Attrs().Flags & net.FlagLoopback
```

insomniacslk replied:
the code is short and readable enough, I agree. However with a dedicated package I just wanted to make sure that users of these libraries get the loopback and non-loopback interfaces properly, and not, for example, by checking that the name is "lo"

Natolumin committed:
netboot: Use dhcp/interfaces instead of netlink

Listing interfaces doesn't need a netlink library at all. The `net`
package provides functions and the dhcp library provides a nice wrapper
to list all non-loopback interfaces

insomniacslk commented:
@Natolumin, I believe this is a sane default, and the new timeout is definitely better. Let's wait for the related PR before merging it

Natolumin commented:
Also I just realized seeing your related PR that I don't need netlink at all here to list the interfaces or even filter them, it's all possible with the standard library `net`
Still conflicted about the added value of the abstraction but now that it's there I'll use it

insomniacslk commented:
> Also I just realized seeing your related PR that I don't need netlink at all here to list the interfaces or even filter them, it's all possible with the standard library net

yes, luckily Go's `net` is much better on that as it doesn't force us to use the Linux-specific netlink :)

> Still conflicted about the added value of the abstraction but now that it's there I'll use it

That PR is not merged yet

insomniacslk commented:
On the value of that PR - this would be used in the dhcpv4 and dhcpv6 package at least, and potentially here, with the goal of making sure that the interfaces are fetched correctly and not my name-matching. I don't have a strong opinion either though, there are pros and cons with both approaches

insomniacslk commented:
@Natolumin the `interfaces` PR is merged in the DHCP library, however feel free to avoid importing it here

insomniacslk commented:
cc @zaolin - this PR will change the default behaviour of `netboot`. It will try DHCP sequentially on every interface, unless a specific NIC is configured via VPD
